src/sse/sse_utils.rs

```diff
@@ -0,0 +1,251 @@
+// |_|  |_|\__,_|\__|_| |_|         |____/_____|_.__/|_|\__|
+//
+
+pub struct Rotate90F32 {
+    //sign_1st: __m128,
+    sign_2nd: __m128,
+    sign_both: __m128,
+}
+
+impl Rotate90F32 {
+    pub fn new(positive: bool) -> Self {
+        // There doesn't seem to be any need for rotating just the first element, but let's keep the code just in case
+        //let sign_1st = unsafe {
+        //    if positive {
+        //        _mm_set_ps(0.0, 0.0, 0.0, -0.0)
+        //    }
+        //    else {
+        //        _mm_set_ps(0.0, 0.0, -0.0, 0.0)
+        //    }
+        //};
+        let sign_2nd = unsafe {
+            if positive {
+                _mm_set_ps(0.0, -0.0, 0.0, 0.0)
+            } else {
+                _mm_set_ps(-0.0, 0.0, 0.0, 0.0)
+            }
+        };
+        let sign_both = unsafe {
+            if positive {
+                _mm_set_ps(0.0, -0.0, 0.0, -0.0)
+            } else {
+                _mm_set_ps(-0.0, 0.0, -0.0, 0.0)
+            }
+        };
+        Self {
+            //sign_1st,
+            sign_2nd,
+            sign_both,
+        }
+    }
+
+    #[inline(always)]
+    pub unsafe fn rotate_2nd(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB4);
+        _mm_xor_ps(temp, self.sign_2nd)
+    }
+
+    // There doesn't seem to be any need for rotating just the first element, but let's keep the code just in case
+    //#[inline(always)]
+    //pub unsafe fn rotate_1st(&self, values: __m128) -> __m128 {
+    //    let temp = _mm_shuffle_ps(values, values, 0xE1);
+    //    _mm_xor_ps(temp, self.sign_1st)
+    //}
+
+    #[inline(always)]
+    pub unsafe fn rotate_both(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB1);
+        _mm_xor_ps(temp, self.sign_both)
+    }
+}
+
+// Pack 1st complex
+// left: r1.re + r1.im, r2.re, r2.im
+// right: l1.re + l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l1.re + l1.im
+#[inline(always)]
+pub unsafe fn pack_1st_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0x44)
+}
+
+// Pack 2nd complex
+// left: r1.re + r1.im, r2.re, r2.im
+// right: l1.re + l1.im, l2.re, l2.im
+// --> r2.re, r2.im, l2.re + l2.im
+#[inline(always)]
+pub unsafe fn pack_2nd_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0xEE)
+}
+
+// Pack 1st and 2nd complex
+// left: r1.re + r1.im, r2.re, r2.im
+// right: l1.re + l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l2.re + l2.im
+#[inline(always)]
+pub unsafe fn pack_1and2_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0xE4)
+}
+
+// Pack 2nd and 1st complex
+// left: r1.re + r1.im, r2.re, r2.im
+// right: l1.re + l1.im, l2.re, l2.im
+// --> r2.re, r2.im, l1.re + l1.im
+#[inline(always)]
+pub unsafe fn pack_2and1_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0x4E)
+}
+
+// Reverse complex
+// values: a.re + a.im, b.re, b.im
```

Comment:
Is the + here intended to be a comma? If so, could you rename this to `reverse_complex_elements_f32` to match the naming scheme of AvxVector?

Reply:
yes and yes :)